File: leetcode/0439-ternary-expression-parser/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public String parseTernary(String expression) {
        Deque<Character> stack = new ArrayDeque<>();

        for (int i = expression.length() - 1; i >= 0; i--) {
            char ch = expression.charAt(i);
            if (!stack.isEmpty() && stack.peek() == '?') {
                stack.pop();                     // discard '?'
                char trueBranch = stack.pop();
                stack.pop();                     // discard ':'
                char falseBranch = stack.pop();
                stack.push(ch == 'T' ? trueBranch : falseBranch);
            } else {
                stack.push(ch);
            }
        }

        return String.valueOf(stack.peek());
    }
}

File: leetcode/0439-ternary-expression-parser/solution.py
class Solution:
    def parseTernary(self, expression: str) -> str:
        stack = []

        for ch in reversed(expression):
            if stack and stack[-1] == '?':
                stack.pop()          # discard '?'
                true_branch = stack.pop()
                stack.pop()          # discard ':'
                false_branch = stack.pop()
                stack.append(true_branch if ch == 'T' else false_branch)
            else:
                stack.append(ch)

        return stack[0]
